campaignion_logcrm.module

```diff
@@ -145,15 +151,52 @@ function campaignion_logcrm_webform_confirm_email_email_confirmed($node, $submis
 function campaignion_logcrm_payment_status_change(Payment $payment, PaymentStatusItem $previous_status_item) {
   $statusChangedToSuccess = $payment->getStatus()->status == PAYMENT_STATUS_SUCCESS && $previous_status_item->status != PAYMENT_STATUS_SUCCESS;
   $hasContextObj = $payment->contextObj instanceof WebformPaymentContext;
-  if (!$statusChangedToSuccess || !$hasContextObj)
+  if (!$statusChangedToSuccess || !$hasContextObj) {
     return;
-  // Send payment event.
-  $e = Event::fromPayment($payment);
-  QueueItem::byData([
-    'type' => 'payment',
-    'id' => $payment->pid,
-    'event' => $e,
-  ])->save();
+  }
+  $exporter = Container::get()->loadService('campaignion_logcrm.payment_exporter');
+  // Queue the payment success event for sending.
+  $queue = Container::get()->loadService('campaignion_logcrm.queue');
+  $queue->addItem('payment', $payment->pid, $exporter->createSuccessEvent($payment));
+}
+
+/**
+ * Implements hook_payment_insert().
+ */
+function campaignion_logcrm_payment_insert(Payment $payment) {
+  return campaignion_logcrm_payment_update($payment);
+}
+
+/**
+ * Implements hook_payment_update().
+ */
+function campaignion_logcrm_payment_update(Payment $payment) {
+  if (!$payment->contextObj instanceof WebformPaymentContext) {
+    return;
+  }
+  $exporter = Container::get()->loadService('campaignion_logcrm.payment_exporter');
+  $queue = Container::get()->loadService('campaignion_logcrm.queue');
+  $last_psiid_handled = $payment->original->psiid_last ?? 0;
+  $pending_after_new = FALSE;
+  $previous_item = NULL;
+  foreach ($payment->statuses as $item) {
+    if (payment_status_is_or_has_ancestor($item->status, PAYMENT_STATUS_NEW)) {
+      $pending_after_new = FALSE;
+    }
+    if (payment_status_is_or_has_ancestor($item->status, PAYMENT_STATUS_PENDING)) {
+      $pending_after_new = TRUE;
+      // Don’t send PENDING statuses themselves.
+      $previous_item = $item;
+      continue;
+    }
+    // Skip anything before PENDING and anything that has already been sent.
+    if (!$pending_after_new || $item->psiid <= $last_psiid_handled) {
+      $previous_item = $item;
+      continue;
+    }
+    $queue->addItem('payment_status_item', $item->psiid, $exporter->createStatusChangeEvent($payment, $previous_item));
+    $previous_item = $item;
+  }
```

Comment:
This requires that *all* payment methods have a pending state, otherwise no status change is ever send. Are you sure they do? It probably does not matter when there is no status change as long as there will be a success status send eventually, but might be an issue when that’s removed? :thinking: 

(Related to https://github.com/moreonion/stripe_payment/pull/43#discussion_r1250456501)

Reply:
Yes, all payment methods do this. `Payment::execute()` sets this status item at the latest. But: Stripe payments are sometimes successful (via JS) before that happens. That’s why I tried to make these payments conform as well. In the notion card for this you can find all the status sequences I have found when looking at live DBs.